src/tools/lib/python/picongpu/utils/find_time.py

```diff
@@ -0,0 +1,163 @@
+        -------
+        A float in seconds.
+        """
+        data_file_path = self.get_data_path()
+
+        # matches floats and scientific floats
+        rg_flt = "([-\+[0-9]+\.[0-9]*[Ee]*[\+-]*[0-9]*)"
+
+        # our UNIT_TIME is scaled to dt
+        dt = np.fromregex(
+            data_file_path,
+            "\s+UNIT_TIME " +
+            rg_flt + "\n",
+            dtype=np.dtype([('dt', 'float')])
+        )
+
+        return dt['dt'][0]
+
+    def get_time(self, iteration):
+        """
+        Find a time in seconds for a given iteration.
+
+        Parameters
+        ----------
+        iteration : (unsigned) integer
+            an iteration
+
+        Returns
+        -------
+        time: float in seconds
+            a matching time
+        """
+        return iteration * self.dt
+
+    def get_iteration(self, t, iterations=None, method='previous'):
+        """
+        Find an iteration for a given time in seconds.
+
+        Parameters
+        ----------
+        t : float
+            time in seconds
+        iterations : np.array of dtype np.uint64
+            an array of iterations to choose from.
+            defaults to full integers
+        methods : string
+            The method how to find a matching iteration.
+            previous : the closest iteration that is <= t (default)
+            closest : the closest iteration to t
+            next : the next iteration that is > t
+
+        Returns
+        -------
+        iteration : (unsigned) integer
+            a matching iteration
+        new_time : float
+            the time at iteration in seconds
+        """
+        if t is None:
+            raise ValueError('The time t needs to be set!')
+
+        implemented_methods = ['previous', 'closest', 'next']
+        if method not in implemented_methods:
+            raise ValueError('The method needs to be one of: {}'
+                             .format(implemented_methods))
+
+        if iterations is None:
+            guess = t / self.dt
+            if method == 'previous':
+                iteration = np.floor(guess)
+            if method == 'closest':
+                iteration = np.round(guess)
+            if method == 'next':
+                iteration = np.ceil(guess)
+            return np.uint64(iteration)
+        else:
+            if type(iterations) is not np.ndarray:
+                raise ValueError('iterations must to be a numpy array!')
+
+        iterations_sorted = np.sort(iterations)
+        times_sorted = iterations_sorted * self.dt
+
+        next_i = np.argmax(times_sorted > t)
+        if next_i > 0:
+            prev_i = next_i - 1
+        else:
+            prev_i = 0
+
+        if method == 'previous':
+            iteration = iterations_sorted[prev_i]
+            if times_sorted[prev_i] > t:
+                raise IndexError('Time t not found in valid range!')
+        if method == 'closest':
+            closest_i = np.abs(times_sorted - t).argmin()
```

Comment:
@ax3l Yes, this is looks more costly on the CPU side, but be aware you are using an interpreter here thus the if clauses above are (more) costly too. Thus in the end you trade multi line python code (**slow**) against probably complicated but fast C++/C/Fortran code which is just a single line in python.

User / code review happiness increases :+1:
compute time decreases :+1:

Okay writing your code in pure C/C++ would be faster 😉